internal/gen-atomicwrapper/main.go

```diff
@@ -0,0 +1,318 @@
+package main
+
+import (
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"go/format"
+	"io"
+	"log"
+	"os"
+	"sort"
+	"strings"
+	"text/template"
+)
+
+func main() {
+	log.SetFlags(0)
+	if err := run(os.Args[1:]); err != nil {
+		log.Fatalf("%+v", err)
+	}
+}
+
+type stringList []string
+
+func (sl *stringList) String() string {
+	return strings.Join(*sl, ",")
+}
+
+func (sl *stringList) Set(s string) error {
+	for _, i := range strings.Split(s, ",") {
+		*sl = append(*sl, strings.TrimSpace(i))
+	}
+	return nil
+}
+
+func run(args []string) error {
+	var opts struct {
+		Name    string
+		Wrapped string
+		Type    string
+
+		Imports      stringList
+		Pack, Unpack string
+
+		CAS  bool
+		Swap bool
+		JSON bool
+
+		File string
+	}
+
+	flag := flag.NewFlagSet("gen-atomicwrapper", flag.ContinueOnError)
+
+	// Required flags
+	flag.StringVar(&opts.Name, "name", "",
+		"name of the generated type (e.g. Duration)")
+	flag.StringVar(&opts.Wrapped, "wrapped", "",
+		"name of the wrapped atomic (e.g. Int64)")
+	flag.StringVar(&opts.Type, "type", "",
+		"name of the type exposed by the atomic (e.g. time.Duration)")
+
+	// Optional flags
+	flag.Var(&opts.Imports, "imports",
+		"comma separated list of imports to add")
+	flag.StringVar(&opts.Pack, "pack", "",
+		"function to transform values with before storage")
+	flag.StringVar(&opts.Unpack, "unpack", "",
+		"function to reverse packing on loading")
+	flag.StringVar(&opts.File, "file", "",
+		"output file path (default: stdout)")
+
+	// Switches for individual methods. Underlying atomics must support
+	// these.
+	flag.BoolVar(&opts.CAS, "cas", false,
+		"generate a `CAS(old, new) bool` method; requires -pack")
+	flag.BoolVar(&opts.Swap, "swap", false,
+		"generate a `Swap(new) old` method; requires -pack and -unpack")
+	flag.BoolVar(&opts.JSON, "json", false,
+		"generate `MarshalJSON/UnmarshJSON` methods")
+
+	if err := flag.Parse(args); err != nil {
+		return err
+	}
+
+	if len(opts.Name) == 0 || len(opts.Wrapped) == 0 || len(opts.Type) == 0 {
+		return errors.New("flags -name, -wrapped, and -type are required")
+	}
+
+	if (len(opts.Pack) == 0) != (len(opts.Unpack) == 0) {
+		return errors.New("either both, or neither of -pack and -unpack must be specified")
+	}
+
+	if opts.CAS && len(opts.Pack) == 0 {
+		return errors.New("flag -cas requires -pack")
+	}
+
+	if opts.Swap && len(opts.Pack) == 0 {
+		return errors.New("flag -swap requires -pack and -unpack")
+	}
+
+	var w io.Writer = os.Stdout
+	if file := opts.File; len(file) > 0 {
+		f, err := os.Create(file)
+		if err != nil {
+			return fmt.Errorf("create %q: %v", file, err)
+		}
+		defer f.Close()
+
+		w = f
+	}
+
+	// Import encoding/json if needed.
+	if opts.JSON {
+		found := false
+		for _, imp := range opts.Imports {
+			if imp == "encoding/json" {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			opts.Imports = append(opts.Imports, "encoding/json")
+		}
+	}
+
+	sort.Strings([]string(opts.Imports))
+
+	data := struct {
```

Comment:
(no action needed) these are all basically the options passed in, should we declare options as a struct, should we just pass that in?

we could add methods on the option struct to convert things like imports from a comma-separated list into `[]string` too

Reply:
Yeah, we can just pass the opts struct in instead of data. I was trying to keep templating and flag parsing separated but that's not necessary for this case I suppose.